sydent/http/servlets/profilereplicationservlet.py

```diff
@@ -0,0 +1,123 @@
+
+MAX_BATCH_SIZE = 200
+
+
+class ProfileReplicationServlet(Resource):
+    isLeaf = True
+
+    def __init__(self, syd):
+        self.sydent = syd
+
+    def _getAllowedHomeservers(self):
+        rawstr = self.sydent.cfg.get('userdir', 'userdir.allowed_homeservers', '')
+        if rawstr == '':
+            return []
+        return [x.strip() for x in rawstr.split(',')]
+
+    def render_POST(self, request):
+        self._async_render_POST(request)
+        return server.NOT_DONE_YET
+
+    @defer.inlineCallbacks
+    def _async_render_POST(self, request):
+        yield
+        try:
+            body = json.load(request.content)
+        except ValueError:
+            request.setResponseCode(400)
+            request.write(json.dumps({'errcode': 'M_BAD_JSON', 'error': 'Malformed JSON'}))
+            request.finish()
+            defer.returnValue(None)
+
+        missing = [k for k in ("batchnum", "batch", "origin_server") if k not in body]
+        if len(missing) > 0:
+            request.setResponseCode(400)
+            msg = "Missing parameters: "+(",".join(missing))
+            request.write(json.dumps({'errcode': 'M_MISSING_PARAMS', 'error': msg}))
+            request.finish()
+            defer.returnValue(None)
+
+        # XXX: verify the sig to auth the source HS
+
+        batchnum = body["batchnum"]
+        batch = body["batch"]
+        origin_server = body["origin_server"]
+
+        allowed_hses = self._getAllowedHomeservers()
+        if not origin_server in allowed_hses:
+            request.setResponseCode(403)
+            request.write(json.dumps({'errcode': 'M_FORBIDDEN', 'error': 'origin server not whitelisted'}))
+            request.finish()
+            defer.returnValue(None)
+
+        profile_store = ProfileStore(self.sydent)
+
+        latest_batch_on_host = profile_store.getLastBatchForOriginServer(origin_server)
+        if latest_batch_on_host is None:
+            latest_batch_on_host = -1
+        if batchnum <= latest_batch_on_host:
+            logger.info("Ignoring batch %d from %s: we already have %d", batchnum, origin_server, latest_batch_on_host)
+            # we already have this batch, thanks
+            request.write(json.dumps({}))
+            request.finish()
+            defer.returnValue(None)
+        elif batchnum == latest_batch_on_host + 1:
+            # good, this is the next batch
+            if len(batch) > MAX_BATCH_SIZE:
+                logger.warn("Host %s sent batch of %s which exceeds max of %d", origin_server, len(batch), MAX_BATCH_SIZE)
+                request.setResponseCode(400)
+                request.write(json.dumps({'errcode': 'M_UNKNOWN', 'error': 'batch size exceeds max of %d' % (MAX_BATCH_SIZE,)}))
+                request.finish()
+                defer.returnValue(None)
+                
+            bad_uids = []
+            for uid, info in batch.items():
+                if 'display_name' not in info or 'avatar_url' not in info:
```

Comment:
I'm slightly surprised we don't just ignore malformed profile updates? Especially as it's presumably quite common for only one of display_name or avatar_url to be updated, so it might feel reasonable to skip the other field?

Reply:
Yeah, it does feel like unchanged fields could be omitted (not that synapse would know how) but this would apply moreso if we had extensible profiles. It would make for some slightly more complex SQL on sydent.